Higher resolution thumbnails

towerofnix opened the pull request:
Resolves #3239.

### Changes

Adds a new addon, "Higher resolution thumbnails"! This includes two primary features:

- Increase resolution of thumbnails embedded across the site. This works dynamically as images are added to the page (e.g. when a reply box is created) and takes into account lazy loading too.
- Fix thumbnails which are initially loaded at a lower resolution than they are actually displayed at (fixing blurry embeds even at 100% zoom and no high DPI/retina).

Here's the settings pane (with default values), including a warning notice about data usage:

![hi-res-thumb-settings](https://user-images.githubusercontent.com/9948030/140863582-418eb044-6a22-419f-9edc-b6459477e901.png)

Probably on most connections/data plans, extra usage from larger thumbnails is hardly an issue, but I thought it still worth pointing out.

### Reason for changes

Things look better! :sparkles: 

Examples (click image to see at original screenshot resolution):

*User avatar in messages, 150% zoom, addon disabled:*

![old-thumb](https://user-images.githubusercontent.com/9948030/140860390-c850b60b-b20e-40e3-974f-6f192a0f1cac.png)

*User avatar in messages, 150% zoom, 150% resolution scaling, "Fix initially low-res thumbnails?" checked:*

![new-thumb](https://user-images.githubusercontent.com/9948030/140860393-ee255987-8b17-4a1f-a698-05e3100015ba.png)

<details><summary>MOAR EXAMPLES</summary>

*User avatar in reply box, 100% zoom, addon disabled:*

![reply-low](https://user-images.githubusercontent.com/9948030/140860419-9c9f23d4-37eb-43aa-9a80-3732605ca482.png)

*User avatar in reply box, 100% zoom, 100% (i.e. no) resolution scaling, "Fix initially low-res thumbnails?" checked:*

![reply-high](https://user-images.githubusercontent.com/9948030/140860428-9b171801-b0c6-4087-a50b-dbcf0eb48e59.png)

*User avatars in forum posts, 130% zoom, addon disabled:*

![forum-low](https://user-images.githubusercontent.com/9948030/140862378-da7d643f-699a-4c12-a8ef-2feedf3f339c.png)

*User avatars in forum posts, 130% zoom, 130% resolution scaling:*

![forum-high](https://user-images.githubusercontent.com/9948030/140862461-7b487f97-4fef-4f3e-af8a-b30526584a8f.png)

*Project thumbnails on user-subpage, 100% zoom, "More row items" addon set to 3 per row, hi-res addon disabled:*

![3row-low](https://user-images.githubusercontent.com/9948030/140862520-1afdda07-2a30-4535-8d7c-f27d44d3940e.png)

*Project thumbnails on user-subpage, 100% zoom, "More row items" addon set to 3 per row, 100% (i.e. no) resolution scaling, "Fix initially low-res thumbnails?" checked:*

![3row-high](https://user-images.githubusercontent.com/9948030/140862581-b86d69cf-f042-465c-a805-3269840f8679.png)

</details>

### Tests

Buncha manual testing :P

Thanks to @FunctionalMetatable in https://github.com/ScratchAddons/ScratchAddons/issues/3239#issuecomment-893326923 for putting together the general userscript outline!

Also, their example has code for thumbnails from uploads.scratch.mit.edu I think, but I'm not exactly sure where that URL pattern appears on the site? Input and further thumbnail locations worth watching out for would be appreciated!

towerofnix commented on addons/hi-res-thumbnails/userscript.js:
```diff
@@ -0,0 +1,77 @@
+        if (scale > 1) {
+          width *= scale;
+          height *= scale;
+        }
```
One point of discussion re: implementation of this resolution fix: maybe it makes sense to *down*scale images which are by default loaded in at a higher resolution than they're actually displayed at? (Before applying the user-specified multiplier, of course.)

The main example I found of this is on the new studio pages, where for some reason thumbnails are loaded at 480x360, even though they're displayed at only about 217x203 in CSS pixels on a 1920p screen. This addon would actually save a little data usage if you set the res setting to 100% and had this fix enabled, then, because it would fetch the images at 217x203 instead of 480x360!

Although apparently, studio thumbnails load a higher res because the rendered size is semi-flexible, i.e. at thinner displays (e.g. by shrinking the window) it'll only show two projects per row, and this can get the thumbnails to around 345x260 CSS pixels. Still less than 480x360, though not by as much. Just the same, it points out a flaw with downscaling to the rendered size: if the image later dynamically *expands*, it'll look lower resolution than the original developers were giving excess dimensions as a way of intentionally allowing for!

towerofnix commented on addons/hi-res-thumbnails/userscript.js:
```diff
@@ -0,0 +1,77 @@
+export default async function ({ addon, console }) {
+  const DEBUG = true;
```
oops

towerofnix commented on addons/hi-res-thumbnails/userscript.js:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  function scaleDimensions(width, height) {
+    const factor = 0.01 * addon.settings.get("multiplier");
```
I decided to make this a setting instead of trying to detect the browser zoom level (or "real" image render size) in some way. Partly because 1) [it's really difficult (possibly impossible) to actually detect that](https://css-tricks.com/can-javascript-detect-the-browsers-zoom-level/), but more importantly because 2) that either wouldn't work with dynamic zooming (i.e. as user zooms further in, image becomes blurry again), or it *would* detect zooming and would cause a TON of slightly-different-resolution image requests as the user adjusts their zoom level.

Leaving it up to the user solves both those problems while also giving them specific control over how big their thumbnails are going to end up, allowing for a maximum "nice" zoom and letting the user decide what they want that to be.

mxmou replied:
I think `devicePixelRatio` and not the zoom level is what you would need, but that still doesn't solve problem 2.

towerofnix replied:
Yeah, `devicePixelRatio` is *theoretically* the solution, but it [isn't consistently supported with regards to zooming](https://css-tricks.com/can-javascript-detect-the-browsers-zoom-level/#there-is-pixel-density).

Another point against dynamically detecting zoom is that users might deliberately zoom further in for the sake of reading *text* more easily (or inspecting smaller layout details)—if a user doesn't care about the image resolution here, currently they can just leave the resolution setting at the max value they deem relevant, instead of loading a bunch of data unnecessarily.

Problem 2 could also cause a lot of lag (replacing image sources frequently), and either we update *all* images (= even more lag and data usage) or we lazily target only images on- or near-screen (= less lag but pop-in as you scroll, and even more code complexity on top of dynamic zoom detection and image updating in the first place).

towerofnix commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+      "type": "integer",
+      "min": 100,
+      "max": 300,
+      "default": 200
```
This is a percentage rather than a decimal (e.g. 2.00x) because I'm not exactly sure decimal settings exist, and because integers are generally nicer for users to play with (citation needed).

It'd be nice to have something like `"step": 10` here (for nicer non-typing controls), but I don't *think* there exists support for that in the current settings viewer?

mxmou replied:
I agree that there should be a `step` option - it would also be helpful for addons that already have percentage settings, such as block transparency.

WorldLanguages replied:
#3860

towerofnix commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "Higher resolution thumbnails",
```
I decided on naming it "Higher resolution thumbnails" instead of "High resolution thumbnails" for a sort of cheeky reason: this way, when you read it, you wonder, "how *much* higher?"—and when you click to find out, it turns out the answer is you get to choose!

The default (200% width & height) is still a sensible value for anyone on a retina screen. If you're the sort of person who browses at a higher than normal zoom level, you probably already understand the specifics of what makes images blurry when you're zoomed in, and will know to set the value to however much you're normally zoomed.

RedGuy12 replied:
yeah, the thing is, now it's a verb, and we want to try and keep addon names as nouns wherever possible

towerofnix commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+      "default": true
+    }
+  ],
+  "tags": ["community"],
```
Maybe more tags would be better? I don't really know where a list of tags or when they're appropriate is, so I'd appreciate review from someone more experienced here :P

mxmou replied:
A slightly outdated list (the "featured" tag is missing) is [here](https://scratchaddons.com/docs/reference/addon-manifest/#tags-array-required). This addon isn't for specific areas of the website and isn't a theme, so I don't think any other tags are necessary.

towerofnix replied:
Thanks for the link, and that makes sense to me!

FunctionalMetatable commented:
> Also, their example has code for thumbnails from uploads.scratch.mit.edu I think, but I'm not exactly sure where that URL pattern appears on the site? Input and further thumbnail locations worth watching out for would be appreciated!

The profile pictures on the "What's happening"  section of the profile uses uploads: https://github.com/LLK/scratch-www/search?q=%22uploads.scratch.mit.edu%22&type=

mxmou commented:
> > Also, their example has code for thumbnails from uploads.scratch.mit.edu I think, but I'm not exactly sure where that URL pattern appears on the site? Input and further thumbnail locations worth watching out for would be appreciated!
> 
> The profile pictures on the "What's happening" section of the profile uses uploads: https://github.com/LLK/scratch-www/search?q=%22uploads.scratch.mit.edu%22&type=

Project and studio thumbnails on the front page are also loaded from `uploads.scratch.mit.edu`, but I don't think those need to be resized at all - they're already loaded at their original size.

towerofnix commented:
> > 
> > The profile pictures on the "What's happening" section of the profile uses uploads: https://github.com/LLK/scratch-www/search?q=%22uploads.scratch.mit.edu%22&type=
> 
> Project and studio thumbnails on the front page are also loaded from `uploads.scratch.mit.edu`, but I don't think those need to be resized at all - they're already loaded at their original size.

Oh, interesting! Thanks both. It seems this is the case for user avatars too—on my homepage right now is [deck26](https://uploads.scratch.mit.edu/users/avatars/1892712.png), [piano_miles](https://uploads.scratch.mit.edu/users/avatars/10494864.png), and [djdolphin](https://uploads.scratch.mit.edu/users/avatars/395764.png), and these are all [already](https://cdn2.scratch.mit.edu/get_image/user/1892712_9999x9999.png) [max](https://cdn2.scratch.mit.edu/get_image/user/10494864_9999x9999.png) [resolution](https://cdn2.scratch.mit.edu/get_image/user/395764_9999x9999.png).

This would be another case for deliberately lowering the resolution re: https://github.com/ScratchAddons/ScratchAddons/pull/3769#discussion_r745289783 (we'd have to manually reconstruct the appropriate cdn2 URL, which @FunctionalMetatable's original code does). But if we aren't for going lower than the intended (base) resolution, because of the problems discussed there, it'd probably just be inconsistent to target these avatar, project, etc images on `uploads` too.

WorldLanguages commented:
Related: "profile page banner" addon also automatically makes the resolution higher for featured project images. Watch out for conflicts?

micahlt commented:
Whoa, just gotta say that this is super cool!  Awesome job 🎉

towerofnix commented:
@WorldLanguages Good point, thanks! I made a small adjustment to that addon's code to support 480x360ifying the thumbnail regardless of its current size. I think it's better to do it this way rather than try to not target the featured project thumbnail from this new addon, only when the *other* addon is detected as enabled (if that's even possible).

RedGuy12 commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "Higher resolution thumbnails",
+  "description": "Replaces thumbnails across the website with higher resolution images. Useful on retina displays or browsing at >100% zoom.",
```
`more than 100%` instead of `>100%`

RedGuy12 commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+  "info": [
+    {
+      "type": "warning",
+      "text": "Higher resolution images require larger filesizes. Take care not to accidentally incur data expenses if your network plan is limited.",
```
worded oddly

RedGuy12 commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+  ],
+  "settings": [
+    {
+      "name": "Dimension Multiplier (%)",
```
it's a bit unclear what this setting does

RedGuy12 commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,42 @@
+      "default": 200
+    },
+    {
+      "name": "Fix initially low-res thumbnails",
```
shouldn't be a setting imo

but if you really want to keep it,
```suggestion
      "name": "Resize initially undersized thumbnails",
```

also, mxmou made an exhaustive list of all sizes of images here fyi: https://github.com/LLK/scratch-www/issues/5598#issuecomment-864118903

towerofnix commented:
Thanks for the feedback @RedGuy12! I updated the strings across the addon.json file, so they should be clearer and easier to understand right away.

I sort of disagree that "Higher resolution thumbnails" is a verb—*higher* isn't a verb, it's an adjective—but "High resolution thumbnails" is punchy and keeping addon names sounding similar is good anyway, so I switched it to that. I also removed the setting for toggling that fix (and instead mentioned it in the addon description)—you're right that toggling it *off* doesn't really serve much purpose.

WorldLanguages commented on addons/hi-res-thumbnails/userscript.js:
```diff
@@ -0,0 +1,68 @@
+export default async function ({ addon, console }) {
+  main: while (true) {
```
Oooh, named loops! :P

towerofnix replied:
Stolen frommm....

@FunctionalMetatable's example code :P https://github.com/ScratchAddons/ScratchAddons/issues/3239#issuecomment-893326923

WorldLanguages commented on addons/hi-res-thumbnails/addon.json:
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "High resolution thumbnails",
+  "description": "Replaces thumbnails across the website with higher resolution images. Useful for Retina displays and browsing at greater than 100% zoom, and scales up thumbnails that are undersized by default.",
```
@towerofnix The description is weird. `Retina` sounds too much like an Apple only thing, maybe "high pixel density displays"? And the last sentence looks out of place, and I'm not sure what it tries to convey.

towerofnix replied:
Yeah, I'm all for "high pixel density displays" :P It's not Apple-exclusive, obviously, but Retina displays afaik are the most like, well-known instance of high DPI displays? I went with it for recognizability—but teaching users to recognize that their Retina display is of the more general (and accurate) class/description "high pixel density display" is worthwhile.

The latter part is about scaling up images that are by default loaded at a *lower* resolution than what they're displayed at, so that they match that display (render) size. So for example, if an avatar is loaded at 38x38 but is displayed at 45x45, it'd automatically replace the dimensions with the larger ones (and then multiply by the user-selected scaling factor).

I'm not sure how to convey that in a simple sentence though, or if it's necessarily a key part of the addon (despite the fact that it's useful even if you don't use the high DPI part of the addon at all). Feedback would be appreciated!

micahlt replied:
How about `Useful for Retina (high pixel density) displays`?  That seems simple enough and it clarifies exactly what you mean by Retina.

towerofnix replied:
Yeah, that'd be good! Another option is `Useful for high pixel density (e.g. Retina) displays`. Same difference really, but it points out explicitly that the more general term is high pixel density.

RedGuy12 reviewed (approved):
wl told me to
